Parsers/ASimNetworkSession/Parsers/ASimNetworkSessionCiscoASA.yml

```diff
@@ -0,0 +1,207 @@
+              NetworkDuration = toint(24 * 60 * totimespan(parsedData[11]) / time(1s)),
+              DstBytes = tolong(parsedData[12]),
+              Reason = coalesce(tostring(reasonParsed[1]), tostring(reasonParsed[3]), tostring(reasonParsed[4]))
+      | lookup EventResultMapping on Reason
+      | extend 
+              SrcUsernameType = iif(isnotempty(SrcUsername),"Windows",""),
+              DstUsernameType = iif(isnotempty(DstUsername),"Windows",""),
+              SimplifiedDeviceAction = DeviceAction,
+              SessionId = NetworkSessionId,
+              EventOriginalResultDetails = iif(isnotempty(EventOriginalResultDetails), strcat(Reason, " - ", EventOriginalResultDetails), EventOriginalResultDetails)
+      | project-away parsedData, reasonParsed, Reason;
+      let all_302020_302021 = allLogs
+      | where DeviceEventClassID in ("302020","302021")
+      | extend SrcUsername = tostring(extract_all(@"\((\S+)\)",Message)[0])      
+      | extend SrcUsernameType = iif(isnotempty(SrcUsername),"Windows","");
+      let all_7_series = allLogs
+      | where DeviceEventClassID in ("710002","710003","710004","710005")
+      | parse Message with * " to " DstInterfaceName ":" *;
+      let remainingLogs = allLogs
+      | where DeviceEventClassID in ("106002", "106007", "106012", "106013", "106017", "106020", "106100");
+      let ExternalInterface = _ASIM_GetWatchlistRaw("NetworkAddresses")
+                                | extend WatchlistItem = parse_json(WatchlistItem)
+                                | where WatchlistItem.Tags has "External"
+                                | distinct tostring(WatchlistItem["Range Name"]);
+      union isfuzzy=true all_106001_alike, all_106010_alike, all_106018, all_106023, all_106023_unparsed, all_106023_41,  all_302013_302015, all_302014_302016, all_302020_302021, all_7_series, remainingLogs
+      | extend 
+              EventStartTime = TimeGenerated,
+              EventEndTime = TimeGenerated,
+              EventVendor = "Cisco",
+              EventProduct = "ASA",
+              EventCount = toint(1),
+              EventType = "NetworkSession",
+              EventSchema = "NetworkSession",
+              EventSchemaVersion = "0.2.3",
+              EventProductVersion = DeviceVersion,
+              EventOriginalSubType = DeviceEventClassID,
+              EventSeverity = LogSeverity,
+              EventOriginalSeverity = OriginalLogSeverity,
+              DvcOriginalAction = DeviceAction,
+              DvcAction = case(SimplifiedDeviceAction in ("Teardown","Built"), "",
+                              SimplifiedDeviceAction),
+              EventMessage = Message,
+              Dvc = Computer,
+              DstPortNumber = coalesce(DstPortNumber, DestinationPort),
```

Comment:
To avoid the coalesce statements, and to use project-rename more, maybe it is better to use two "mega" unions, one for the parsed events, and one for the events you need to parse. 

I suspect this is also useful for the filtering parameters in the vim parser.